server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -0,0 +1,774 @@
+import org.opensearch.common.lucene.search.AutomatonQueries;
+import org.opensearch.core.xcontent.DeprecationHandler;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.common.xcontent.JsonToStringXContentParser;
+import org.opensearch.index.analysis.IndexAnalyzers;
+import org.opensearch.index.analysis.NamedAnalyzer;
+import org.opensearch.index.fielddata.IndexFieldData;
+import org.opensearch.index.fielddata.plain.SortedSetOrdinalsIndexFieldData;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.index.query.QueryShardException;
+import org.opensearch.search.aggregations.support.CoreValuesSourceType;
+import org.opensearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import static org.opensearch.search.SearchService.ALLOW_EXPENSIVE_QUERIES;
+
+/**
+ * A field mapper for flat_objects.
+ * This mapper accepts JSON object and treat as string fields in one index.
+ * @opensearch.internal
+ */
+public final class FlatObjectFieldMapper extends DynamicKeyFieldMapper {
+
+    public static final String CONTENT_TYPE = "flat_object";
+    private static final String VALUE_AND_PATH_SUFFIX = "._valueAndPath";
+    private static final String VALUE_SUFFIX = "._value";
+    private static final String DOT_SYMBOL = ".";
+    private static final String EQUAL_SYMBOL = "=";
+
+    /**
+     * In flat_object field mapper, field type is similar to keyword field type
+     * Cannot be tokenized, can OmitNorms, and can setIndexOption.
+     * @opensearch.internal
+     */
+    public static class Defaults {
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.freeze();
+        }
+
+    }
+
+    @Override
+    public MappedFieldType keyedFieldType(String key) {
+        return new FlatObjectFieldType(this.name() + DOT_SYMBOL + key);
+    }
+
+    /**
+     * FlatObjectFieldType is the parent field type.
+     */
+    public static class FlatObjectField extends Field {
+
+        public FlatObjectField(String field, BytesRef term, FieldType ft) {
+            super(field, term, ft);
+        }
+
+    }
+
+    private static FlatObjectFieldMapper toType(FieldMapper in) {
+        return (FlatObjectFieldMapper) in;
+    }
+
+    /**
+     * The builder for the flat_object field mapper using default parameters
+     * @opensearch.internal
+     */
+    public static class Builder extends FieldMapper.Builder<Builder> {
+
+        private final IndexAnalyzers indexAnalyzers;
```

Comment:
`indexAnalyzers` are also not used, please remove them

Reply:
indexAnalyzers is used in the constructor of FlatObjectFieldMapper [here](https://github.com/opensearch-project/OpenSearch/blob/e1c17f0cd4ca2d6fbefffc8961cbd34d3a1ece59/server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java#L526)

Reply:
but it is not used anywhere, you could remove it from constructor

Reply:
make sense, since we are not taking IndexAnalyzer from user input, We set IndexAnalyzer inside the fieldType so we can remove it completely from the FlatObjectFieldMapper Builder.

Reply:
removed in the new commit. Thanks @reta !

Reply:
all checks are passed, do you have other concerns that I did not address？